core/src/main/java/org/jruby/runtime/load/LibrarySearcher.java

```diff
@@ -359,20 +478,58 @@ public void clear() {
             featurePaths.clear();
         }
 
-        final CharSequence key;
-        final CopyOnWriteArrayList<IRubyObject> featurePaths;
+        final StringWrapper key;
+        final List<IRubyObject> featurePaths;
     }
 
-    protected void addSingleFeatureToIndex(String key, IRubyObject featurePath) {
-        Map<String, Feature> featuresIndex = this.loadedFeaturesIndex;
+    private final ThreadLocal<StringWrapper> keyWrapper = ThreadLocal.withInitial(() -> new StringWrapper(null, 0, 0));
```

Comment:
is this really worth the thread-local, do we allocate that many string instances on load service lookups?